### PR TITLE
Serialbox compare

### PR DESCRIPTION
**Pull request by thfabian**
## Technical description

This PR ports the **compare** tool to Serialbox2. The tool was overhauled and rewritten in Python3 using the Python module of serialbox.

### compare.py

All command-line arguments and functionality of the old compare tool has been ported to the new one. The new updated usage can be found below:
```
usage: compare.py [-h] [--version] [-v] [--no-color] [-m NUM] [-s REGEX]
                  [-t TOL] [-q] [-i START[:END]] [-j START[:END]]
                  [-k START[:END]] [-l START[:END]]
                  FILE_1 FILE_2

Compares Serialbox fields in FILE_1 and FILE_2. FILE_1 and FILE_2 can be .dat
or .json files, but need to be of the same type. When they are .json files,
all fields are compared. When they are .dat files, the field name has to be
the last part of the file name, separated by '_'.

positional arguments:
  FILE_1                Path to a field file (.dat) or archive (.json)
  FILE_2                Path to a field file (.dat) or archive (.json)

optional arguments:
  -h, --help            show this help message and exit
  --version             show version information and exit
  -v, --verbose         enable verbose logging
  --no-color            disabled colored output (default: False)
  -m NUM, --max-errors NUM
                        Report up to 'NUM' errors per field (default: 10)
  -s REGEX, --savepoint-filter REGEX
                        Only compare fields of savepoints whose name matches
                        REGEX. An empty string will match every savepoint
                        (default: ".*-in")
  -t TOL, --tolerance TOL
                        set the tolerance used for comparison to 'TOL'
                        (default : 1e-12)
  -q, --info-only       Only compare field meta-info (no data comparison)
  -i START[:END]        only compare the i dimension 'START' or if 'END' is
                        supplied compare in the range ['START', 'END']
  -j START[:END]        only compare the j dimension 'START' or if 'END' is
                        supplied compare in the range ['START', 'END']
  -k START[:END]        only compare the k dimension 'START' or if 'END' is
                        supplied compare in the range ['START', 'END']
  -l START[:END]        only compare the l dimension 'START' or if 'END' is
                        supplied compare in the range ['START', 'END']
```

#### Extensions & Changes
- The output was completely overhauled and now uses a colored output following googles GTest style (color can be disabled with `--no-color`)
- Error reporting was improved. Individual errors can now be printed (by default the first 10 errors, incuding their values and index in the field, are displayed which can be adjusted with the command-line flag `-m`)
- A new command line option `-s REGEX` has been added which allows to only compare the fields of savepoints whose name matches `REGEX`. By default the regex is set to `.*-in` which is identical to the hardcoded behavior of the old compare tool.
- The tool is now installed in `python/compare/compare.py` as opposed to the previous install location `bin/`.

### Additional fixes
 - **pp_ser** was missing the license header.
 - The version of the Python serialbox module was still at 2.0.1. Now it's correctly at 2.1.1.
 - The python script was missing some quotes in the paths.

@clementval  It would be appreciated to test to the tool in the wild to provide some Feedback.

### CODE REVIEW

File: src/serialbox-python/compare/compare.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
+#!/usr/bin/python3
```
Comment (clementval) [inline src/serialbox-python/compare/compare.py]: On Kesch this fails
Works with -> `#!/usr/bin/env python3`

File: src/serialbox-python/compare/compare.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+
+from __future__ import print_function
+
+from sys import exit, stderr, version_info
+
+# Check Python version
+if version_info < (3, 4):
+    from platform import python_version
+
+    print("compare: error: compare requires atleast python 3.4 (detected %s)" % python_version(),
```
Comment (clementval) [inline src/serialbox-python/compare/compare.py]: Typo: at least with a space between at and least

File: src/serialbox-python/pp_ser/pp_ser.py [hunk 3]
```diff
@@ -1,6 +1,13 @@
 #!/usr/bin/env python
-# This file is released under terms of BSD license`
-# See LICENSE.txt for more information
+# -*- coding: utf-8 -*-
```
Comment (clementval) [inline src/serialbox-python/pp_ser/pp_ser.py]: Was this header really necessary?

File: src/serialbox-python/compare/compare.py (hunk 1 above)
Comment (thfabian) [reply src/serialbox-python/compare/compare.py]: done

File: src/serialbox-python/compare/compare.py (hunk 2 above)
Comment (thfabian) [reply src/serialbox-python/compare/compare.py]: done

File: src/serialbox-python/pp_ser/pp_ser.py (hunk 3 above)
Comment (thfabian) [reply src/serialbox-python/pp_ser/pp_ser.py]: Well I would aim for consistency.. either remove all the headers if you don't like it or put atleast the same on every file -- I don't care which way you want to go ;)

Comment (clementval) [reply src/serialbox-python/pp_ser/pp_ser.py]: I was just teasing your ASCII art skills

**Comment by clementval** [comment]
@thfabian Would it be easy to just have a mode where metedata difference would just give a warning instead of a fail? Just I think we will see lots of comparison fails with old archive.

**Comment by clementval** [comment]
Otherwise lgtm

**Comment by thfabian** [comment]
@clementval Yes I added a flag which prints a warning on difference.

**Comment by pspoerri** [comment]
What's the rule for the default name if I pass a folder? Can you clarify the error a bit when comparing two folders and the metainfo file is not found?

```
> ./compare.py /scratch/jenkins/data/cosmo-prerelease/double/cosmo1_test3/ /scratch/jenkins/data/cosmo-prerelease/float/cosmo1_test3/
Traceback (most recent call last):
  File "./compare.py", line 404, in <module>
    exit(main())
  File "./compare.py", line 394, in main
    serializer_1, field_1 = make_serializer_and_extract_field(path_1)
  File "./compare.py", line 131, in make_serializer_and_extract_field
    serializer = ser.Serializer(ser.OpenModeKind.Read, directory, prefix)
  File "/scratch-shared/meteoswiss/scratch/spoerrip/proj_serialbox2/serialbox2/build_gcc_cray/install/python/compare/../serialbox/serializer.py", line 257, in __init__
    archivestr)
  File "/scratch-shared/meteoswiss/scratch/spoerrip/proj_serialbox2/serialbox2/build_gcc_cray/install/python/compare/../serialbox/error.py", line 57, in invoke
    raise SerialboxError(error_message.decode())
serialbox.error.SerialboxError: cannot create Serializer: MetaData-cosmo1.json not found in "/scratch/jenkins/data/cosmo-prerelease/double"
```

File: test/serialbox-python/run_python_tests.sh.cmake [hunk 4]
```diff
@@ -17,12 +17,12 @@
 #
 # Export paths
 #
-export PYTHONPATH=$PYTHONPATH:${SERIALOBX_PYTHON_MODULE}:${SERIALOBX_PYTHON_MODULE}/sdb
+export PYTHONPATH="$PYTHONPATH:${SERIALOBX_PYTHON_MODULE}:${SERIALOBX_PYTHON_MODULE}/sdb"
```
Comment (clementval) [inline test/serialbox-python/run_python_tests.sh.cmake]: Just saw a typo in the SERIALOBX_PYTHON_MODULE var

**Comment by clementval** [comment]
@thfabian I see a flag `-q` for comparing only metadata but can I compare only data and NOT the metadata?

**Comment by clementval** [comment]
@thfabian There is an exception if the savepoint is only present in one archive. This is often the case with input data. I think it should just skip this savepoint. 

```
MetaData-radiation-standalone_rank0.json 
[==========] Set-up serializers.
[          ]   serializer_1 = '/scratch-shared/meteoswiss/scratch/clementv/serialbox_migration/physics-standalone/radiation/claw/run' (prefix 'Field_rank0')
[          ]   serializer_2 = '/scratch-shared/meteoswiss/scratch/clementv/serialbox_migration/physics-standalone/radiation/claw/run' (prefix 'radiation-standalone_rank0')
[----------] Savepoint verif-radiation.input {"iteration": 0}
[ RUN      ] rad_w_so
Traceback (most recent call last):
  File "./compare", line 403, in <module>
    exit(main())
  File "./compare", line 399, in main
    return compare([serializer_1, serializer_2], field_1, dim_bounds, args.info_only)
  File "./compare", line 280, in compare
    if not compare_infos(serializers, field):
  File "./compare", line 137, in compare_infos
    info_1, info_2 = serializer_1.get_field_metainfo(field), serializer_2.get_field_metainfo(field)
  File "/project/c14/install/kesch/serialbox2/gnu/python/serialbox/serializer.py", line 599, in get_field_metainfo
    fieldstr))
  File "/project/c14/install/kesch/serialbox2/gnu/python/serialbox/fieldmetainfo.py", line 106, in __init__
    typeid = c_int(type)
TypeError: an integer is required (got type NoneType)
```

**Comment by thfabian** [comment]
Yeah I frogot about catching some exceptions, my bad. I will change this stuff

File: test/serialbox-python/run_python_tests.sh.cmake (hunk 4 above)
Comment (thfabian) [reply test/serialbox-python/run_python_tests.sh.cmake]: Nice catch! `sdb` is abandoned anyway though ;)

**Comment by clementval** [comment]
@havogt This can be merged from my side!

**Pull request merged**